instant-distance-py/src/lib.rs

```diff
@@ -24,41 +26,84 @@ fn instant_distance_py(_: Python, m: &PyModule) -> PyResult<()> {
     m.add_class::<Search>()?;
     m.add_class::<Hnsw>()?;
     m.add_class::<HnswMap>()?;
+    m.add_class::<DistanceMetric>()?;
     Ok(())
 }
 
+#[pyclass]
+#[derive(Copy, Clone, Default)]
+enum DistanceMetric {
+    #[default]
+    Euclid,
+    Cosine,
+    DotProduct,
+}
+
 #[pyclass]
 struct HnswMap {
-    inner: instant_distance::HnswMap<FloatArray, MapValue>,
+    inner: HnswMapWithMetric,
+}
+
+#[derive(Deserialize, Serialize)]
+enum HnswMapWithMetric {
+    Euclid(instant_distance::HnswMap<FloatArray<EuclidMetric>, MapValue>),
+    Cosine(instant_distance::HnswMap<FloatArray<CosineMetric>, MapValue>),
+    DotProduct(instant_distance::HnswMap<FloatArray<DotProductMetric>, MapValue>),
+}
+
+// Helper macro for dispatching to inner implementation
+macro_rules! impl_for_each_hnsw_with_metric {
+    ($type:ident, $instance:expr, $inner:ident, $($tokens:tt)+) => {
+        match $instance {
+            $type::Euclid($inner) => {
+                $($tokens)+
+            }
+            $type::Cosine($inner) => {
+                $($tokens)+
+            }
+            $type::DotProduct($inner) => {
+                $($tokens)+
+            }
+        }
+    };
```

Comment:
Style nit: let's move this down below the `Hnsw`.

Reply:
Since it's a macro it has to be defined before the first use (in `HnswMap::search`).